### ISSUE TITLE
[ENH] sktime-pywatts integration - `__call__` dunder design

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#2654]
Part of [pywatts integration design questions](https://github.com/alan-turing-institute/sktime/issues/2653).

`pywatts` has a nice `keras`-like `__call__` dunder syntax for creating pipelines.

It is, however, not entirely clear:

* which design we should follow - single layer (modules and estimators "fused") or dual layer (pywatts modules vs sktime estimators)
* whether `__call__` should live directly in `sktime`, or in `pywatts`. If not in `sktime`, how could it possibly work after import?

Some preliminary thoughts are in this document: https://github.com/sktime/community-org/blob/main/meetups/previous_meetings/20220506-pywatts.md

FYI @aiwalter

**Comment by fkiraly** [comment]
FYI, this may interest you, @miraep8, due to connection with dunders/pipelines.

**Comment by benHeid** [comment]
Concerning the second bullet point, I just had an idea to make the `__call__` dunder import-dependent. However, I suppose this solution is a bit hacky... and also confusing to users because we probably could not customize the exception message if the users try to use the `__call__` under without having installed pyWATTS.

``` python
# CASE 1: The bases contain the JointAPIBase since the package some_non_installed is not installed

class OtherBases():
    def print_1(self):
        print(1)

class JointAPIBase():
    def print_2(self):
        print(2)

try:
    import os # some_non_installed
    bases = [JointAPIBase, OtherBases]
except ImportError as error:
    bases = [OtherBases]

class BaseSKtime(*bases):
    pass

bar = BaseSKtime()
bar.print_1()
try:
    bar.print_2()
except Exception:
    pass


# CASE 2: The bases does not contain the JointAPIBase since the package some_non_installed is not installed


try:
    import some_non_installed
    bases = [JointAPIBase, OtherBases]
except ImportError as error:
    bases = [OtherBases]

class BaseSKtime(*bases):
    pass

bar = BaseSKtime()
bar.print_1()
try:
    bar.print_2()
except Exception:
    print("Print2 does not exist")
    pass
```

`some_non_installed ` and `os`  would be replaced by pyWATTS and `print_2` by `__call__`.  This would allow that depending on whether the user has installed pyWATTS the sktime module has a `__call__` dunder or not.

**Comment by fkiraly** [comment]
some incomplete opinions on pros/cons

* modifying structure of estimators is against user expectation in sklearn. I would go with (d) = no
* sklearn has (d) = no, by clone - this is the user expectation. pywatts does by reference.
    * Benedikt: we may want to use pretrained components in a pipeline
    * FK: would do this with a `Pretrained` wrapper, which does nothing in `fit` but retrieves
    * pywatts does that via `computation_mode` argument, in the call dunder. If value is `"transform"`, fitting is skipped. Similar `"refit"`
    * FK: `sktime` has some similar concepts - `FitInTransform`, or `RefitEvery` - these are not call args but wrappers
* if (a) = `GraphPipeline`, we could lose internal `StepInformation`, but it probably still makes sense to have two layers

**Comment by SMEISEN** [comment]
- (a) __call__ return: StepInformation vs GraphPipeline
    - `StepInformation` is an Object with the Attributes `Step` and `Pipeline`
       connecting steps is performed like the current pyWATTS implementation but `StepInformation.Pipeline.train(data)` enables training the (sub-)pipeline

**Comment by benHeid** [comment]
* (a) 
   * Drawbacks return Graphpipeline:
     * Each pipeline has only one output or last step (This could probably depend on the choice on decision point (d)).
   * Benefits return Graphpipeline:
     * More intuitive to use?
* (c) Input Subsetting
   * Persional Opinion: I would prefer to use a dummy object (`Column()`) or a string, because:
     * this is probably an operation which is often used. Thus, using a ColumnSelectorTransformer could be cause a lot of code in the pipeline definition and could make the pipeline defintion confusing
     * the square bracket dunder (`__getattr__`) is used in pandas to select a column of a dataframe. Transferred to the pipeline, it would mean that with the square brackets one can select values the pipeline stands for and for me this are the output data.
     * Using a string or a dummy object is in some way analog to the API of Keras.
* (c)

**Comment by miraep8** [comment]
(c) regarding subsetting:
-  Personally, I would agree with @benHeid that it makes sense to use a Dummy variable for input subsetting.  I just like the idea of separating the specification of a model from the data it is being used to model. 
- I think there is an additional advantage here of making it easier to specify and share models across problems as well. 

(d) Here is an idea:
- A potentially interesting way to use "build-by-clone" but still get the advantages of a pre-trained model could be to add in the ability to save and share fitted parameters in each of our pipelines and models.  This is an approach that obviously has some traction already in the Neural Nets world (especially with regard to Transformers) but I think we could benefit from it here as well.  Ultimately I think it could help make it easier for us to facilitate the "load pre-trained parameters" functionality, and for people to share models and results with each other as well! Could also be a helpful practice to put into place now since it seems like `sktime` will soon have more deep learning components and something like parameter saving and sharing seems like it would be extra useful in that context.

**Comment by fkiraly** [comment]
@miraep8, could you be more concrete of how your implementation to "load fitted parameters" would work?
Were you thinking of an extra method, a feature of the call dunder, or sth else?

**Comment by fkiraly** [comment]
design study of `FittedModel` here: https://github.com/alan-turing-institute/sktime/pull/2903

**Comment by miraep8** [comment]
I checked out the `FittedModel` PR and I think this approach seems reasonable. 

I guess one question is - I suppose the 'model' that we are fitting/reloading will be a `GraphPipeline` estimator/object? So then the question for me is - do we need to create a new class that is this `FittedModel`, or can we just use the `GraphPipeline` class, but have the option of generating this from unpacking a `pickle`?

**Comment by miraep8** [comment]
Ahhh, I see.  The advantage of having the `FittedModel` as a separate class` is that on `fit` we can skip fitting the parameters.  Which does not happen if we just treat it as a normal `GraphPipeline` object.  

I guess this could get us to a separate question of - if a part of the model is already fitted, we could skip re-fitting it, and instead only update the fitted parameters if `update` is called. 

But I understand the design choice better now of making a new class.

**Comment by fkiraly** [comment]
Note from @kalebphipps

* simplicity is important for users - e.g., dunders, square brackets, dummy objects
* putting things "under the hood" if more intuitive is good. Should not require more complex workflows like "build stepinformation and then wrap it" etc

**Comment by fkiraly** [comment]
@miraep8:

* prefers "everything is `GraphPipeline`"
* sees that "build model" might be necessary and interim objects that have "multiple ends"
* could `GraphPipeline` also have "multiple ends" in scope and construct only when final object has a "type"
* do `GraphPipeline` need to be constructed in "forward" direction?
* maybe need separate assembly/build process

**Comment by fkiraly** [comment]
here's a suggestion on this question:

> whether `__call__` should live directly in `sktime`, or in `pywatts`. If not in `sktime`, how could it possibly work after import?

I think a good solution would be the following:
* `GraphPipeline` lives in `pywatts`, but is `sktime` compatible
* `pywatts` is an `sktime` soft dependency, localized to `__call__`
* `sktime` is a `pywatts` soft dependency, for base class (parent of `GraphPipeline`?) and testing
* testing for `GraphPipeline` uses `sktime` `check_estimator`
* `sktime` base classes implement high-level `__call__` logic, i.e., constructing the `GraphPipeline`
* `sktime` base classes import `pywatts` `GraphPipeline` internally in `__call__`, errors are raised via `_check_soft_dependency`. This is similar to how the other dunders import a specific compositor class, e.g., `*` of `BaseTransformer` importing the `TransformerPipeline`.

**Comment by benHeid** [comment]
We decided to use a design without the`__dunder__`. Thus I am closing this issue. 

Feel free to reopen it, if you disagree.
